### PR TITLE
#2423 New embedded WebExtension add-on

### PR DESCRIPTION
**Pull request by lmorchard**
To build: `cd addon && npm install && npm start`

This is all still very much in progress, but it's close enough in shape to start getting some critique. Some remaining TODOs to get done before considering this not WIP:

- [x] "New" badge text when new experiments & news updates found? (#2325, #2419)
- [x] Ensure existing build-and-sign in Jenkins works
- [x] Ensure l10n for add-on gets carried over as necessary
- [x] Ensure add-on build in CircleCI works
- [x] Metrics pings from experiments
- [x] Metrics pings on extension enable / disable
- [x] Open channels for known list of experiments (mainly for webextensions)
- [x] Set prefs to enable Telemetry with preservation & restore of original settings on uninstall
- [x] Replace `addon/` with `addon-bootstrapped/` (temporary working directory)
- [x] Squash / reorg my dirty in-progress commits
- [x] <strike>Variant testing support? (don't think we use this, probably worth omitting)</strike>
- [x] <strike>Messaging channel to website to enable add-on installation</strike>
- [x] environment switch via about:config
- [x] toolbar button
- [x] installation detection in web frontend via mozAddonManager
- [x] ensure `navigator.testpilotAddon` properties are populated for `about:home` and `mozilla.com` sites
- [x] nsIObserver events for legacy add-on metrics pings
- [x] BroadcastChannel messages for WebExtension metrics pings
- [x] periodically fetch experiments list & news updates from site

I should flesh out the README, but here are some TL;DRs on how I'm building it:

* webextension side should coordinate & drive everything with an eye toward discarding privileged APIs when possible
* bootstrap side is mainly intended as a service provider of privileged APIs for webextension side
* trying to organize everything into purpose-specific modules to concentrate concerns
* dropped redux, introduced pubsub - things look less like state management with side effects and more like stateless message handling with small data caches
* lost flow types & tests because I stumbled & experimented my way into how things work now, and neither were helping me along the way - should backfill
* using webpack & babel, but mainly for the module system - Firefox supports advanced JS features and doesn't need (much) code transformation. (Very tempted to dump webpack in favor of rollup & npm scripts like I did on Snooze Tabs.)

### CODE REVIEW

**Comment by lmorchard** [comment]
Also, of interest for trying this thing out: I exercised the nsIObserver events & BroadcastChannel events by temporarily installing the bootstrapped & webextension add-ons from the `docs/examples` directory. Should see some Browser Toolbox log messages as those add-ons make contact with this add-on for metrics

**Comment by lmorchard** [comment]
Also also: Very much welcoming any TODOs that I might have missed from staring at the current add-on code over the past month.

**Comment by lmorchard** [comment]
More progress - if you have the three current experiments installed (i.e. snooze tabs, minvid, and containers), you should see log chatter in Browser Toolbox about metrics pings. Need to double check that all the right stuff is getting sent along to the respective services though

**Comment by lmorchard** [comment]
Okay. So I think I have the build process all working like the current add-on. Might work a bit over the weekend on getting linting and tests working. 

But, otherwise, I think I've covered off on the current add-on's functionality. Which means I've probably missed something.

**Comment by lmorchard** [comment]
Updated to re-implement "new" badge text on experiment updates

File: addon/src/bootstrap.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    PubSub.clearAllSubscriptions();
+  } catch (err) {
+    log('shutdown error', err);
+  }
+}
+
+async function setupDebug() {
+  if (!debug) return;
+  ['bootstrap', 'webExtension'].forEach(root =>
+    PubSub.subscribe(root, (message, data) => log('pubsub', message, data))
```
Comment (fzzzy) [inline addon/src/bootstrap.js:63]: Logging root might be really useful for debugging.

Comment (lmorchard) [reply addon/src/bootstrap.js:63]: I don't think PubSub-JS supports that, but could look into monkey patching it

Comment (fzzzy) [reply addon/src/bootstrap.js:63]: Ok, I thought log could just take arbitrary arguments.

Comment (lmorchard) [reply addon/src/bootstrap.js:63]: Oh wait, you mean literally the variable `root` - I thought you meant the root of pubsub messages :) Yeah, no, log totally takes arbitrary arguments. But, `message` will always start with the value of `root` as a prefix

**Pull request merged**